File: src/external/Bass/c/dsp/dsp.c
/***************************************************************************
 dsp.c/h/rc - Copyright (c) 2002-2008 (: JOBnik! :) [Arthur Aminov, ISRAEL]
                                                    [http://www.jobnik.org]
                                                    [   bass_fx@jobnik.org]
 
 BASS_FX dsp fx in action
 * Imports: bass.lib, bass_fx.lib
            kernel32.lib, user32.lib, comdlg32.lib, gdi32.lib
***************************************************************************/

#include <windows.h>
#include <commctrl.h>
#include <stdio.h>
#include "bass.h"
#include "bass_fx.h"
#include "dsp.h"

HWND win=NULL;
HINSTANCE inst;

HFONT Font;					// font handle
DWORD chan;					// a handle
DWORD floatable;			// floating-point channel support?
HFX fxEQ, fxPhaser;			// dsp fx handles
float oldfreq;				// old sample rate

#define MESS(id,m,w,l) SendDlgItemMessage(win,id,m,(WPARAM)w,(LPARAM)l)

OPENFILENAME ofn;
char path[MAX_PATH];

// display error dialogs
void Error(const char *es)
{
	char mes[200];
	sprintf(mes,"%s\n\n(error code: %d)",es,BASS_ErrorGetCode());
	MessageBox(win,mes,"Error",MB_ICONEXCLAMATION);
}

// update dsp eq
void UpdateFX(int b)
{
	BASS_BFX_PEAKEQ eq;
	int v = MESS(IDC_SLDEQ1+b,TBM_GETPOS,0,0);

	eq.lBand = b;	// get values of the selected band
	BASS_FXGetParameters(fxEQ, &eq);
		eq.fGain = (float)(10 - v);
	BASS_FXSetParameters(fxEQ, &eq);
}

// set dsp eq
void SetDSP_EQ(float fGain, float fBandwidth, float fQ, float fCenter_Bass, float fCenter_Mid, float fCenter_Treble)
{
	BASS_BFX_PEAKEQ eq;

	// set peaking equalizer effect with no bands
	fxEQ=BASS_ChannelSetFX(chan, BASS_FX_BFX_PEAKEQ,0);

	eq.fGain=fGain;
	eq.fQ=fQ;
	eq.fBandwidth=fBandwidth;
	eq.lChannel=BASS_BFX_CHANALL;

	// create 1st band for bass
	eq.lBand=0;
	eq.fCenter=fCenter_Bass;
	BASS_FXSetParameters(fxEQ, &eq);

	// create 2nd band for mid
	eq.lBand=1;
	eq.fCenter=fCenter_Mid;
	BASS_FXSetParameters(fxEQ, &eq);

	// create 3rd band for treble
	eq.lBand=2;
	eq.fCenter=fCenter_Treble;
	BASS_FXSetParameters(fxEQ, &eq);

	// update dsp eq
	UpdateFX(0);
	UpdateFX(1);
	UpdateFX(2);
}

// get the file name from the file path
char *GetFileName(char *filepath)
{
	unsigned char slash_location;
	filepath = strrev(filepath);
	slash_location = strchr(filepath, '\\') - filepath;
	return (strrev(filepath) + strlen(filepath) - slash_location);
}

BOOL CALLBACK dialogproc(HWND h,UINT m,WPARAM w,LPARAM l)
{
	DWORD a=0;
	float freq;
	char c[30];

	BASS_BFX_PEAKEQ eq;		// dsp peaking equalizer
	BASS_BFX_PHASER phs;	// dsp phaser

	switch (m) {
		case WM_COMMAND:
			switch (LOWORD(w)) {
				case ID_OPEN:
					{
						char file[MAX_PATH]="";
						ofn.lpstrFilter="playable files\0*.mo3;*.xm;*.mod;*.s3m;*.it;*.mtm;*.mp3;*.mp2;*.mp1;*.ogg;*.wav;*.aif\0All files\0*.*\0\0";
						ofn.lpstrFile=file;
						if (GetOpenFileName(&ofn)) {
							memcpy(path,file,ofn.nFileOffset);
							path[ofn.nFileOffset-1]=0;

							// free previous dsp effects & handles
							BASS_StreamFree(chan);	// free stream
							BASS_MusicFree(chan);	// free music

							if(!(chan=BASS_StreamCreateFile(FALSE, file, 0, 0, BASS_SAMPLE_LOOP|floatable))&&
								!(chan=BASS_MusicLoad(FALSE, file, 0, 0, BASS_MUSIC_LOOP|BASS_MUSIC_RAMP|floatable,0))){
								// not a WAV/MP3 or MOD
								MESS(ID_OPEN,WM_SETTEXT,0,"click here to open a file && play it...");
								Error("Selected file couldnt be loaded!");
								break;
							}

							// update the Button to show the loaded file name
							MESS(ID_OPEN,WM_SETTEXT,0,GetFileName(file));

							// set dsp effects
							SendMessage(win,WM_COMMAND,IDC_CHKEQ,l);
							SendMessage(win,WM_COMMAND,IDC_CHKPHS,l);

							// get current sample rate
							BASS_ChannelGetAttribute(chan, BASS_ATTRIB_FREQ, &freq);
							oldfreq = freq;

							// set the dx sample rate & view
							MESS(IDC_DXRATE,TBM_SETRANGEMAX,0,(long)(freq * 1.3f));
							MESS(IDC_DXRATE,TBM_SETRANGEMIN,0,(long)(freq * 0.7f));
							MESS(IDC_DXRATE,TBM_SETPOS,TRUE,(long)freq);
							MESS(IDC_DXRATE,TBM_SETPAGESIZE,0,(long)(freq * 0.01f));	// by 1%

							sprintf(c,"DirectX Samplerate = %dHz", (long)freq);
							MESS(IDC_SDXRATE,WM_SETTEXT,0,c);

							// play it!
							BASS_ChannelPlay(chan, FALSE);
						}
					}
					return 1;

				case IDC_CHKEQ:
					if (MESS(IDC_CHKEQ,BM_GETCHECK,0,0))
						SetDSP_EQ(0.0f, 2.5f, 0.0f, 125.0f, 1000.0f, 8000.0f);
					else
						BASS_ChannelRemoveFX(chan, fxEQ);
				return 1;

				case IDC_CHKPHS:
					if(MESS(IDC_CHKPHS,BM_GETCHECK,0,0)){
						fxPhaser=BASS_ChannelSetFX(chan, BASS_FX_BFX_PHASER,0);

						BASS_FXGetParameters(fxPhaser, &phs);
							phs.fWetMix = (float)MESS(IDC_WETMIX,TBM_GETPOS,0,0) / 1000.0f;
							phs.fDryMix = (float)MESS(IDC_DRYMIX,TBM_GETPOS,0,0) / 1000.0f;
							phs.fFeedback = (float)MESS(IDC_FEEDBACK,TBM_GETPOS,0,0) / 1000.0f;
							phs.fRate = (float)MESS(IDC_RATE,TBM_GETPOS,0,0) / 10.0f;
							phs.fRange = (float)MESS(IDC_RANGE,TBM_GETPOS,0,0) / 10.0f;
							phs.fFreq = (float)MESS(IDC_FREQ,TBM_GETPOS,0,0) / 10.0f;
						BASS_FXSetParameters(fxPhaser, &phs);
					}else
						BASS_ChannelRemoveFX(chan, fxPhaser);
				return 1;

			}
			return 1;

		case WM_VSCROLL:
			if(l){
				UpdateFX(GetDlgCtrlID((HWND)l)-IDC_SLDEQ1);
			}
		return 1;

		case WM_HSCROLL:
			if(!BASS_ChannelIsActive(chan)) break;

			switch (GetDlgCtrlID((HWND)l)) {
				case IDC_DXRATE:
					BASS_ChannelSetAttribute(chan, BASS_ATTRIB_FREQ, (float)MESS(IDC_DXRATE, TBM_GETPOS, 0, 0));

					sprintf(c,"DirectX Samplerate = %dHz", MESS(IDC_DXRATE, TBM_GETPOS, 0, 0));
					MESS(IDC_SDXRATE,WM_SETTEXT,0,c);

					// update all bands fCenters after changing samplerate
					{
						int i;
						for(i=0;i<3;i++){
							eq.lBand = i;
							BASS_FXGetParameters(fxEQ, &eq);
								eq.fCenter = eq.fCenter * (float)MESS(IDC_DXRATE, TBM_GETPOS, 0, 0) / oldfreq;
							BASS_FXSetParameters(fxEQ, &eq);
						}
						oldfreq = (float)MESS(IDC_DXRATE, TBM_GETPOS, 0, 0);
					}
				break;

				case IDC_DRYMIX:
				case IDC_WETMIX:
				case IDC_FEEDBACK:
				case IDC_RATE:
				case IDC_RANGE:
				case IDC_FREQ:
					BASS_FXGetParameters(fxPhaser, &phs);
						phs.fWetMix = (float)MESS(IDC_WETMIX,TBM_GETPOS,0,0) / 1000.0f;
						phs.fDryMix = (float)MESS(IDC_DRYMIX,TBM_GETPOS,0,0) / 1000.0f;
						phs.fFeedback = (float)MESS(IDC_FEEDBACK,TBM_GETPOS,0,0) / 1000.0f;
						phs.fRate = (float)MESS(IDC_RATE,TBM_GETPOS,0,0) / 10.0f;
						phs.fRange = (float)MESS(IDC_RANGE,TBM_GETPOS,0,0) / 10.0f;
						phs.fFreq = (float)MESS(IDC_FREQ,TBM_GETPOS,0,0) / 10.0f;
					BASS_FXSetParameters(fxPhaser, &phs);
			}
			return 1;

		case WM_INITDIALOG:
			win=h;
			GetCurrentDirectory(MAX_PATH,path);
			memset(&ofn,0,sizeof(ofn));
			ofn.lStructSize=sizeof(ofn);
			ofn.hwndOwner=h;
			ofn.hInstance=inst;
			ofn.nMaxFile=MAX_PATH;
			ofn.Flags=OFN_HIDEREADONLY|OFN_EXPLORER;

			// enable floating-point DSP
			BASS_SetConfig(BASS_CONFIG_FLOATDSP, TRUE);

			// setup output - default device, 44100hz, stereo, 16 bits
			if (!BASS_Init(-1,44100,0,win,NULL)) {
				Error("Can't initialize device");
				DestroyWindow(win);
				return 1;
			}

			// check for floating-point capability
			floatable = BASS_StreamCreate(44100, 2, BASS_SAMPLE_FLOAT, 0, 0);
			if (floatable) {
				BASS_StreamFree(floatable);  //woohoo!
			    floatable = BASS_SAMPLE_FLOAT;
			}

			// initialize dsp eq sliders
			MESS(IDC_SLDEQ1,TBM_SETRANGE,FALSE,MAKELONG(0,20));
			MESS(IDC_SLDEQ1,TBM_SETPOS,TRUE,10);

			MESS(IDC_SLDEQ2,TBM_SETRANGE,FALSE,MAKELONG(0,20));
			MESS(IDC_SLDEQ2,TBM_SETPOS,TRUE,10);

			MESS(IDC_SLDEQ3,TBM_SETRANGE,FALSE,MAKELONG(0,20));
			MESS(IDC_SLDEQ3,TBM_SETPOS,TRUE,10);

			// dx rate
			MESS(IDC_DXRATE,TBM_SETRANGEMAX,0,(long)(44100.0f*1.3f));
			MESS(IDC_DXRATE,TBM_SETRANGEMIN,0,(long)(44100.0f*0.7f));
			MESS(IDC_DXRATE,TBM_SETPOS,TRUE,44100);
			MESS(IDC_DXRATE,TBM_SETPAGESIZE,0,(long)(44100.0f*0.01f));	// by 1%

			// DryMix
			MESS(IDC_DRYMIX,TBM_SETRANGE,0,MAKELONG(-2000,2000));
			MESS(IDC_DRYMIX,TBM_SETPOS,TRUE,-999);

			// WetMix
			MESS(IDC_WETMIX,TBM_SETRANGE,0,MAKELONG(-2000,2000));
			MESS(IDC_WETMIX,TBM_SETPOS,TRUE,999);

			// Feedback
			MESS(IDC_FEEDBACK,TBM_SETRANGE,0,MAKELONG(-1000,1000));
			MESS(IDC_FEEDBACK,TBM_SETPOS,TRUE,-60);

			// Rate
			MESS(IDC_RATE,TBM_SETRANGE,0,MAKELONG(0,100));
			MESS(IDC_RATE,TBM_SETPOS,TRUE,2);

			// Range
			MESS(IDC_RANGE,TBM_SETRANGE,0,MAKELONG(0,100));
			MESS(IDC_RANGE,TBM_SETPOS,TRUE,60);

			// Freq
			MESS(IDC_FREQ,TBM_SETRANGE,0,MAKELONG(0,10000));
			MESS(IDC_FREQ,TBM_SETPOS,TRUE,1000);

			Font=CreateFont(-12,0,0,0,FW_BOLD,FALSE,FALSE,FALSE,ANSI_CHARSET,OUT_STRING_PRECIS,
				            CLIP_STROKE_PRECIS,DRAFT_QUALITY,DEFAULT_PITCH | FF_DONTCARE,
							"MS Sans Serif");
			// set the font for check boxes
			MESS(IDC_CHKEQ, WM_SETFONT, Font, TRUE);
			MESS(IDC_CHKPHS,WM_SETFONT, Font, TRUE);
		return 1;

		case WM_CLOSE:
			EndDialog(h,0);
			return 0;
		break;
	}
	return 0;
}

int PASCAL WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance,LPSTR lpCmdLine, int nCmdShow)
{
	inst=hInstance;

	// check the correct BASS was loaded
	if (HIWORD(BASS_GetVersion())!=BASSVERSION) {
		MessageBox(0,"An incorrect version of BASS.DLL was loaded (2.4 is required)","Incorrect BASS.DLL",MB_ICONERROR);
		return 1;
	}

	// check the correct BASS_FX was loaded
	if (HIWORD(BASS_FX_GetVersion())!=BASSVERSION) {
		MessageBox(0,"An incorrect version of BASS_FX.DLL was loaded (2.4 is required)","Incorrect BASS_FX.DLL",MB_ICONERROR);
		return 1;
	}

	DialogBox(inst,(char*)1000,0,&dialogproc);

	BASS_Free();

	DeleteObject(Font);

	return 0;
}

File: src/external/Bass/delphi/DSP/dsp.pas
{========================================================================
 dsp.pas - Copyright (c) 2003-2008 (: JOBnik! :) [Arthur Aminov, ISRAEL]
                                                 [http://www.jobnik.org]
                                                 [   bass_fx@jobnik.org]

 BASS_FX dsp fx in action
 * Requires: BASS 2.4 (available @ www.un4seen.com)
========================================================================}

unit dsp;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  StdCtrls, ComCtrls, BASS, BASS_FX;

type
  TfrmDSP = class(TForm)
    btnOpen: TButton;
    od: TOpenDialog;
    GroupBox: TGroupBox;
    lbl1KHz: TLabel;
    lbl8KHz: TLabel;
    lblDryMix: TLabel;
    lblWetMix: TLabel;
    lblFeedback: TLabel;
    lblDX: TLabel;
    lbl125Hz: TLabel;
    lblRate: TLabel;
    vsEQBASS: TScrollBar;
    vsEQMID: TScrollBar;
    vsEQTREBLE: TScrollBar;
    chkPhaser: TCheckBox;
    tbDryMix: TTrackBar;
    tbWetMix: TTrackBar;
    tbFeedback: TTrackBar;
    tbDXRate: TTrackBar;
    chkEqualizer: TCheckBox;
    tbRate: TTrackBar;
    lblRange: TLabel;
    tbRange: TTrackBar;
    lblFreq: TLabel;
    tbFreq: TTrackBar;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure btnOpenClick(Sender: TObject);
    procedure lblDXClick(Sender: TObject);
    procedure tbDXRateChange(Sender: TObject);
    procedure chkEqualizerClick(Sender: TObject);
    procedure vsEQBASSChange(Sender: TObject);
    procedure vsEQMIDChange(Sender: TObject);
    procedure vsEQTREBLEChange(Sender: TObject);
    procedure chkPhaserClick(Sender: TObject);
    procedure tbDryMixChange(Sender: TObject);
    procedure tbWetMixChange(Sender: TObject);
    procedure tbFeedbackChange(Sender: TObject);
    procedure Error(msg: string);
    procedure UpdateEQ(b, pos: integer);
    procedure UpdatePHS();
  private
    { Private declarations }
    chan: DWORD;               // a handle
    freq: FLOAT;               // sample rate
    oldfreq: FLOAT;            // old sample rate
    floatable: DWORD;	       // floating-point channel support?
    fxEQ : DWORD;              // dsp peaking eq handle
    fxPhaser : DWORD;          // dsp phaser handle
    eq  : BASS_BFX_PEAKEQ;   // dsp peaking equalizer
    phs : BASS_BFX_PHASER;   // dsp phaser
  public
    { Public declarations }
  end;

var
  frmDSP: TfrmDSP;

implementation

{$R *.DFM}

// display error dialogs
procedure TFrmDSP.Error(msg: string);
var s: string;
begin
    s := msg + #13#10 + #13#10 + '(error code: ' + IntToStr(BASS_ErrorGetCode) + ')';

  MessageBox(handle, PChar(s), 'Error', MB_ICONEXCLAMATION);
end;

procedure TfrmDSP.FormCreate(Sender: TObject);
begin
  // check the correct BASS was loaded
  If (HiWord(BASS_GetVersion()) <> BASSVERSION) then begin
    application.MessageBox('An incorrect version of BASS.DLL was loaded (2.4 is required)', 'Incorrect BASS.DLL', MB_ICONERROR or MB_OK);
    application.Terminate;
  end;

  // check the correct BASS_FX was loaded
  If (HiWord(BASS_FX_GetVersion()) <> BASSVERSION) then begin
    application.MessageBox('An incorrect version of BASS_FX.DLL was loaded (2.4 is required)', 'Incorrect BASS_FX.DLL', MB_ICONERROR or MB_OK);
    application.Terminate;
  end;

  // enable floating-point DSP
  BASS_SetConfig(BASS_CONFIG_FLOATDSP, 1);

  // setup output - default device, 44100hz, stereo, 16 bits
  If BASS_Init(-1, 44100, 0, frmDSP.handle, nil) = false then begin
    Error('Couldnt Initialize Digital Output');
    application.Terminate;
  end;

  // check for floating-point capability
  floatable := BASS_StreamCreate(44100, 2, BASS_SAMPLE_FLOAT, nil, nil);
  if (floatable > 0) then begin
    BASS_StreamFree(floatable);  //woohoo!
    floatable := BASS_SAMPLE_FLOAT;
  end;

  freq := 44100;
end;

procedure TfrmDSP.FormDestroy(Sender: TObject);
begin
 BASS_Free;    // free BASS & BASS_FX
end;

procedure TfrmDSP.btnOpenClick(Sender: TObject);
begin
  od.Filename := '';
  od.Filter :=  'Playable files|*.mo3;*.xm;*.mod;*.s3m;*.it;*.mtm;*.mp3;*.mp2;*.mp1;*.ogg;*.wav;*.aif|All files|*.*';
  If od.Execute = false then exit;

  // free previous streams/music/dsps handles
  BASS_StreamFree(chan);
  BASS_MusicFree(chan);

  // check for a stream
  chan := BASS_StreamCreateFile(FALSE, pChar(od.filename), 0, 0, BASS_SAMPLE_LOOP Or floatable);

  // check for MOD
  If chan = 0 Then chan := BASS_MusicLoad(FALSE, pChar(od.filename), 0, 0, BASS_MUSIC_LOOP Or BASS_MUSIC_RAMP Or floatable, 0);

  If chan = 0 then begin
    btnOpen.Caption := 'click here to open a file && play it...';
    Error('Selected file couldnt be loaded!');
    Exit;
  end;

  // set dsp effects
  chkEqualizer.OnClick(Self);
  chkPhaser.OnClick(Self);

  // set the dx sample rate
  BASS_ChannelGetAttribute(chan, BASS_ATTRIB_FREQ, freq);
  oldfreq := freq;

  with tbDXRate do begin
    min := 0;    // to avoit min > max
    max := 1;    // to avoid max < min
    max := round(freq * 1.3);
    min := round(freq * 0.7);
    pagesize := round(freq * 0.01);      // by 1%
  end;

  // update the button to show the loaded filename
  btnOpen.Caption := ExtractFileName(pChar(od.Filename));

  // update lblDX Hint
  lblDX.hint := 'Click to set DX Samplerate to ' + FloatToStr(freq) + 'Hz';

  // it must be one of them :)
  BASS_ChannelPlay(chan, FALSE);

  // update the dx sample rate to normal
  lblDX.OnClick(self);
end;

procedure TfrmDSP.lblDXClick(Sender: TObject);
begin
  tbDXRate.Position := Round(freq);
  tbDXRate.OnChange(self);
end;

procedure TfrmDSP.tbDXRateChange(Sender: TObject);
var i: DWORD;
begin
  If BASS_ChannelIsActive(chan) = 0 Then Exit;

  BASS_ChannelSetAttribute(chan, BASS_ATTRIB_FREQ, tbDXRate.position);

  lblDX.Caption := 'DirectX Samplerate = ' + IntToStr(tbDXRate.position) + 'Hz';

  // update all bands fCenters after changing samplerate
  for i := 0 to 2 do begin
    eq.lBand := i;
    BASS_FXGetParameters(fxEQ, @eq);
      eq.fCenter := eq.fCenter * tbDXRate.position / oldfreq;
    BASS_FXSetParameters(fxEQ, @eq);
  end;

  oldfreq := tbDXRate.position;
end;

procedure TfrmDSP.chkEqualizerClick(Sender: TObject);
begin
  If chkEqualizer.checked Then begin

    // set peaking equalizer effect with no bands
    fxEQ := BASS_ChannelSetFX(chan, BASS_FX_BFX_PEAKEQ, 0);

    with eq do begin
      fBandwidth := 2.5;
      FQ := 0;
      fGain := 0;
      lChannel := BASS_BFX_CHANALL;

      // create 1st band for bass
      eq.lBand := 0;
      eq.fCenter := 125;
      BASS_FXSetParameters(fxEQ, @eq);

      // create 2nd band for mid
      eq.lBand := 1;
      eq.fCenter := 1000;
      BASS_FXSetParameters(fxEQ, @eq);

      // create 3rd band for treble
      eq.lBand := 2;
      eq.fCenter := 8000;
      BASS_FXSetParameters(fxEQ, @eq);
    end;

    // update dsp eq
    vsEQBASS.OnChange(Self);
    vsEQMID.OnChange(Self);
    vsEQTREBLE.OnChange(Self);
  end
  Else
    BASS_ChannelRemoveFX(chan, fxEQ);
end;

procedure TfrmDSP.vsEQBASSChange(Sender: TObject);
begin
  UpdateEQ(0, vsEQBASS.Position);
end;

procedure TfrmDSP.vsEQMIDChange(Sender: TObject);
begin
  UpdateEQ(1, vsEQMID.Position);
end;

procedure TfrmDSP.vsEQTREBLEChange(Sender: TObject);
begin
  UpdateEQ(2, vsEQTREBLE.Position);
end;

procedure TfrmDSP.chkPhaserClick(Sender: TObject);
begin
  If chkPhaser.checked Then begin
    fxPhaser := BASS_ChannelSetFX(chan, BASS_FX_BFX_PHASER, 1);
      tbDryMix.OnChange(Self);
      tbWetMix.OnChange(Self);
      tbFeedback.OnChange(Self);
      tbRate.OnChange(Self);
      tbRange.OnChange(Self);
      tbFreq.OnChange(Self);
  end
  Else
    BASS_ChannelRemoveFX(chan, fxPhaser);
end;

procedure TfrmDSP.tbDryMixChange(Sender: TObject);
begin
  UpdatePHS;
end;

procedure TfrmDSP.tbWetMixChange(Sender: TObject);
begin
  UpdatePHS;
end;

procedure TfrmDSP.tbFeedbackChange(Sender: TObject);
begin
  UpdatePHS;
end;

// update the dsp eq fx
procedure TfrmDSP.UpdateEQ(b, pos: integer);
begin
  eq.lBand := b;    // get b band values

  BASS_FXGetParameters(fxEQ, @eq);
    eq.fGain := 10 - pos;
  BASS_FXSetParameters(fxEQ, @eq);
end;

// update the dsp phaser fx
procedure TfrmDSP.UpdatePHS();
begin
  BASS_FXGetParameters(fxPhaser, @phs);
    phs.fDryMix := tbDryMix.Position / 1000;
    phs.fWetMix := tbWetMix.Position / 1000;
    phs.fFeedback := tbFeedback.Position / 1000;
    phs.fRate := tbRate.Position / 10;
    phs.fRange := tbRange.Position / 10;
    phs.fFreq := tbFreq.Position / 10;
  BASS_FXSetParameters(fxPhaser, @phs);
end;

end.
